File: python/private_key/5_limit_order.py
# python 5_limit_order.py  — GTC BUY 2 ticks below best bid → cancel
import os
import requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import (
    OrderArgsV2, PartialCreateOrderOptions, OrderType,
    BalanceAllowanceParams, AssetType, OrderPayload,
)

load_dotenv()

PRIVATE_KEY = os.environ["PRIVATE_KEY"]
CLOB_HOST   = os.environ.get("CLOB_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL     = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID    = "73106"
PUSD        = Web3.to_checksum_address("0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB")
ERC20_ABI   = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
                "inputs": [{"name": "account", "type": "address"}],
                "outputs": [{"name": "", "type": "uint256"}]}]

w3  = Web3(Web3.HTTPProvider(RPC_URL))
eoa = w3.eth.account.from_key("0x" + PRIVATE_KEY).address

temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=0)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=0)
print(f"EOA     : {eoa}")
print(f"API key : {creds.api_key} ✓")

# Sync on-chain balance/allowance to CLOB server cache
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.COLLATERAL))

event     = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id   = event["markets"][0]["conditionId"]
info      = client.get_clob_market_info(cond_id)
token_id  = next((t["t"] for t in info.get("t", []) if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)

book  = client.get_order_book(token_id)
bids  = sorted(book["bids"], key=lambda b: float(b["price"]), reverse=True)
price = float(bids[2]["price"])  # 2 ticks below best bid

balance = w3.eth.contract(address=PUSD, abi=ERC20_ABI).functions.balanceOf(eoa).call() / 1e6
size    = max(round((balance - 1) / price, 2), 0)
print(f"Balance : ${balance:.4f} pUSD")
print(f"Placing GTC BUY: {size} shares @ ${price}")

resp     = client.create_and_post_order(
    OrderArgsV2(token_id=token_id, price=price, size=size, side="BUY"),
    PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk),
    OrderType.GTC,
)
order_id = resp.get("orderID") if isinstance(resp, dict) else resp.orderID
print(f"✓ Order placed : {order_id}  status: {resp.get('status') if isinstance(resp, dict) else resp.status}")

cancel   = client.cancel_order(OrderPayload(orderID=order_id))
canceled = cancel.get("canceled", []) if isinstance(cancel, dict) else []
print(f"✓ Cancelled" if order_id in canceled else f"Cancel resp: {cancel}")

File: typescript/private_key/5_limit_order.ts
// npm run 5:limit  — GTC BUY 2 ticks below best bid → cancel
import "dotenv/config";
import { createWalletClient, createPublicClient, http, erc20Abi, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType } from "@polymarket/clob-client-v2";

const CLOB_HOST = process.env.CLOB_BASE_URL ?? "https://clob-v2.polymarket.com";
const PUSD      = "0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB" as `0x${string}`;
const EVENT_ID  = "73106";

const account      = privateKeyToAccount(`0x${process.env.PRIVATE_KEY!}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http() });
const publicClient = createPublicClient({ chain: polygon, transport: http() });

// Init client
const tempClient = new ClobClient({ host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0 });
const creds      = await (tempClient as any).createOrDeriveApiKey();
const client     = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});

// Market info
const event    = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId   = event.markets?.[0]?.conditionId as string;
const info     = await (client as any).getClobMarketInfo(condId);
const tokens   = (info?.t ?? []) as { t: string; o: string }[];
const tokenId  = tokens.find(t => t.o === "Yes")?.t ?? "";
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk  = info?.nr ?? false;

// Order book — pick 2 ticks below best bid
const book  = await (client as any).getOrderBook(tokenId);
const bids  = [...(book?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
const price = parseFloat(bids[2].price);

// pUSD balance
const rawBal  = await publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [account.address] });
const balance = parseFloat(formatUnits(rawBal, 6));
const size    = Math.floor(((balance - 1) / price) * 100) / 100;

console.log(`Placing GTC BUY: ${size} shares @ $${price}`);

const resp = await client.createAndPostOrder(
  { tokenID: tokenId, price, size, side: Side.BUY },
  { tickSize, negRisk },
  OrderType.GTC,
);
console.log(`Order ID : ${resp.orderID}  status: ${resp.status}`);

// Cancel (cleanup)
const cancel = await client.cancelOrder({ orderID: resp.orderID! });
console.log(cancel?.canceled?.includes(resp.orderID!) ? "✓ Cancelled" : JSON.stringify(cancel));
